### PR TITLE
Add kubevirt platform

### PR DESCRIPTION
**Pull request by ravidbro**
Add KubeVirt as a platform for running OpenShift nodes as VMs.

More details at https://github.com/openshift/enhancements/pull/417

### CODE REVIEW

**Review by cgwalters — CHANGES_REQUESTED** [review]
I love the idea.  So meta!

But I really strongly think before we try to land this in 4.7 we need to invest a lot in de-duping and rationalizing the platform-specific code in the MCO.  We simply cannot carry *another* copy of the kubelet configs for example.

**Comment by cgwalters** [comment]
Some of that is in https://github.com/openshift/machine-config-operator/pull/2079

**Comment by ravidbro** [comment]
> Some of that is in #2079

I agree.
When I created the PR I saw that #2079 wasn't approved yet and also that it doesn't contain all the files (e.g. kubelet, keepalived).
I think I will wait for the templates de-dup PR and then I will rebase on it.

**Comment by ravidbro** [comment]
@cgwalters I changed the PR to have only the bump of openshift/api library and the manifests of the bootstrap based on #2071 .
I will add the required files for the template after #2079 will be merged.

File: templates/common/kubevirt/units/qemu-guest-agent.service.yaml
Comment (cgwalters) [inline templates/common/kubevirt/units/qemu-guest-agent.service.yaml:26]: This must be fixed before this can merge.

**Comment by cgwalters** [comment]
/hold
For the above at the minimum.
Can you please add a link in the PR and commit message to https://github.com/openshift/enhancements/pull/417 at least?

File: templates/common/kubevirt/units/qemu-guest-agent.service.yaml
Comment (ravidbro) [reply templates/common/kubevirt/units/qemu-guest-agent.service.yaml]: For now, I removed the guest agent, we can work around without that

**Comment by ravidbro** [comment]
> /hold
> For the above at the minimum.
> Can you please add a link in the PR and commit message to [openshift/enhancements#417](https://github.com/openshift/enhancements/pull/417) at least?

I added a reference to the enhancement both in the PR and in the commit message.

File: pkg/controller/template/test_data/controller_config_kubevirt.yaml
Comment (rgolangh) [inline pkg/controller/template/test_data/controller_config_kubevirt.yaml:33]: nodeDNSIP is redundant

**Review by rgolangh — COMMENTED** [review]
small nit, otherwise lgtm

File: pkg/controller/template/test_data/controller_config_kubevirt.yaml
Comment (ravidbro) [reply pkg/controller/template/test_data/controller_config_kubevirt.yaml]: Thanks, I removed it

**Comment by ravidbro** [comment]
@cgwalters I fixed the issues you were pointing on.
Can you approve it or is there something else?

**Comment by cgwalters** [comment]
/hold cancel
Since https://github.com/openshift/machine-config-operator/pull/2098#discussion_r516534826 is addressed.
Let's get one other MCO member to do the approve.

**Comment by sinnykumari** [comment]
overall looks fine.

One question before approving the PR: I see that linked https://github.com/openshift/enhancements/pull/417 hasn't been approved/signed-off by anyone yet. Shouldn't we wait for approval or having some sort of agreement on enhancement before merging this?

**Comment by ravidbro** [comment]
Thanks @sinnykumari 
The enhancement wasn't merged yet but it was already agreed with the PMs and the stakeholders and we are aiming for dev preview in 4.7

**Comment by sinnykumari** [comment]
> Thanks @sinnykumari
> The enhancement wasn't merged yet but it was already agreed with the PMs and the stakeholders and we are aiming for dev preview in 4.7

Thanks for clarifying :)
/approve

**Comment by darkmuggle** [comment]
With the concerns previously addressed, this looks good to me. Since I'm new to the MCO table, I'll leave the LGTM to @sinnykumari.

File: install/0000_80_machine-config-operator_00_namespace.yaml
Comment (runcom) [inline install/0000_80_machine-config-operator_00_namespace.yaml:60]: we're not going with this - we're decoupling everything not MCO that does leverage the MCO in this way. There are BZs around this as well. We'd need to discuss this further - the full rationale is that we need to find a way for my team to own everything in the MCO repo w/o saying "this isn't our team's stuff". This goes against that

**Comment by runcom** [comment]
While this PR is in line with what we historically did see https://github.com/openshift/machine-config-operator/pull/2098/files#r519658424 - we'd need to find a better way to handle such changes and work towards decoupling what we have already (not add more).
/hold

File: install/0000_80_machine-config-operator_00_namespace.yaml
Comment (ravidbro) [reply install/0000_80_machine-config-operator_00_namespace.yaml:60]: It makes a lot of sense to decouple them, and I'll be happy to implement it differently.
Is there already another way to achieve that?
We aim to support the new platform as dev-preview in 4.7, and the KNI-networking is a must for that.
Can you refer me to the work that is done around that, maybe tickets/BZs.
If the issue is mainly creating a new resource (the namespace), I believe we can create it in the platform-specific area in the installer, but adding cases in the existing switch-cases should probably need to be done here.

Comment (runcom) [reply install/0000_80_machine-config-operator_00_namespace.yaml:60]: we're unfortunately not there yet as far as a new design/architecture goes - time wise I think landing the templates is ok'ish as long as we decouple the namespace. This is a cross-team discussion that we haven't fully started yet but we will asap

Comment (ravidbro) [reply install/0000_80_machine-config-operator_00_namespace.yaml]: Thank you.
I removed the namespace manifest from the 'install' dir

Comment (ravidbro) [reply install/0000_80_machine-config-operator_00_namespace.yaml]: The decoupling work sounds like a good approach, and since we are maintaining a platform (KubeVirt) for IPI, I'll be happy to be involved or at least updated.
Are there already any tickets that we can follow?

Comment (ravidbro) [reply install/0000_80_machine-config-operator_00_namespace.yaml]: @runcom I tested this branch with an installer that creates the namespace in its manifests, and the installation finished successfully. The cluster is functioning as expected with all the KNI components.
Can we cancel the /hold now?

**Comment by runcom** [comment]
@yuqi-zhang @sinnykumari @kikisdeliveryservice ptal as well 

/hold cancel

**Comment by kikisdeliveryservice** [comment]
It feels premature to merge this seeing that the enhancement is not approved - missing both approved & lgtm) as well as a test plan: https://github.com/openshift/enhancements/pull/417

Given the above I don't feel like this should be merged until the above is satisfied.

/hold

**Comment by kikisdeliveryservice** [comment]
Q: is installer intending to merge this PR for 4.7? https://github.com/openshift/installer/pull/4350

**Comment by ravidbro** [comment]
yes, the installer PR is intended to merge for 4.7.
We need also this (MCO) PR to go in since we are committed for a customer to have dev-preview of this feature in 4.7.
The enhancement is not merged but there is already an agreement on that, you can see @crawford comment

**Comment by chenyosef** [comment]
> It feels premature to merge this seeing that the enhancement is not approved - missing both approved & lgtm) as well as a test plan & dev preview criteria: [openshift/enhancements#417](https://github.com/openshift/enhancements/pull/417)

@kikisdeliveryservice Thanks for your review. As for the test plan - our plan is to use the same testing framework as other platforms, i.e. using OpenShift CI (you can see WIP in https://github.com/openshift/release/pull/13547, https://github.com/openshift/release/pull/12765, https://github.com/openshift/release/pull/13165), including the conformance testing of the tenant cluster). We will update the enhancement to reflect that (like @ravidbro mentioned, the enhancement itself was already discussed and approved by both PMs and @crawford, even though wasn't merged yet). 
More over, as for the timeline - yes, we planned to integrate this to 4.7, and already did so in several other components (https://github.com/openshift/cloud-credential-operator/pull/260, https://github.com/openshift/cluster-api-provider-kubevirt, https://github.com/openshift/release/blob/0ac2cccd3e33a748523cb484a20ea27307a3f8b6/core-services/image-mirroring/openshift/mapping_origin_4_7#L140)
> 
> Given the above I don't feel like this should be merged until the above is satisfied.
> 
> /hold
Given the above can you please consider removing the hold ? 

Thanks,
Chen

**Comment by kikisdeliveryservice** [comment]
hi @chenyosef & @ravidbro !

Thanks for the info.

2 Qs:
1) what is being delivered for dev preview/alpha? (the enhancement doesn't say)
2) when do you intend on having any basic e2e on this? I'd assume that once all of this merges, we'd be able to reopen https://github.com/openshift/release/pull/13165 and get that working soon after (so we wouldnt be waiting multiple releases right)?

**Comment by ravidbro** [comment]
1. The dev preview will have fully functional environment installed as IPI (except of automating csi driver install)
The reason for the dev preview is that for this version it is tested only by us and the CI and not by QE.
2. We expect the CI to land in the next few weeks

**Comment by kikisdeliveryservice** [comment]
>     1. The dev preview will have fully functional environment installed as IPI (except of automating csi driver install)
>        The reason for the dev preview is that for this version it is tested only by us and the CI and not by QE.
> 
>     2. We expect the CI to land in the next few weeks

Great!  Thanks for the info!  Our gcp-op is required to run and pass before merging this PR but we are having issues with that test for the last week that will be solved once https://github.com/openshift/machine-config-operator/pull/2229 merges. To avoid unnecessary retests (it will def fail rn), I'll remove the hold tomorrow.

**Comment by ravidbro** [comment]
> > ```
> > 1. The dev preview will have fully functional environment installed as IPI (except of automating csi driver install)
> >    The reason for the dev preview is that for this version it is tested only by us and the CI and not by QE.
> > 
> > 2. We expect the CI to land in the next few weeks
> > ```
> 
> Great! Thanks for the info! Our gcp-op is required to run and pass before merging this PR but we are having issues with that test for the last week that will be solved once #2229 merges. To avoid unnecessary retests (it will def fail rn), I'll remove the hold tomorrow.
> 
> Add: In the meantime it looks like this PR needs a rebase anyway.

Thank you for the review.
I see that the fix was merged and I already rebased my branch to include all the fixes from master.
I will highly appreciate it if you can cancel the hold at this point.

**Comment by ravidbro** [comment]
I see that all tests passed but the 'okd' one which fails constantly on "toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit" because docker started with their rate-limiting.

@kikisdeliveryservice can you cancel the hold?
highly appreciated if you can help us to meet our goal.

**Comment by kikisdeliveryservice** [comment]
I'm in PST so I just am going thru things.

/hold cancel

**Pull request merged**
